### ISSUE TITLE
Mitigating Spectre attacks

### ISSUE THREAD

**Issue by jfoote** [stoklund/cretonne#220]
Hello,

You are probably well aware, but some mainstream compilers are emitting [retpolines](https://support.google.com/faqs/answer/7625886) to help mitigate [Spectre variant 2 attacks](https://spectreattack.com/spectre.pdf). Do you have any plans to add  a similar capability to the cretonne code generator (and/or do you think it makes sense for cretonne to do this sort of thing)?

Thanks,
Jon

/cc @tyler @pchickey

**Comment by stoklund** [comment]
Yes, I expect we will need mitigations for all variants of Spectre. All of these attacks depend on a high-resolution timer being available to malicious code, so a first-level mitigation is to make sure that the available timer resolution is low enough to make the possible exfiltration bandwidth impractical.

Beyond that, I expect the following areas to be affected in Cretonne:

## Heap bounds checking

Heaps that have 4 GB of guard pages should not be affected, but heap styles that use less virtual address space will generate the kind of bounds checking code that can be exploited to read the full 4 GB of address space following the heap base. We can implement different styles of mitigation:

- If the heap maximum size is small, we can use a constant *index mask* to limit the range of memory that can be read by speculative code. See [the WebKit blog post about index masking](view-source:https://webkit.org/blog/8048/what-spectre-and-meltdown-mean-for-webkit/).
- A dynamic index mask may also be appropriate in some cases.
- We can construct a mask using `setCC` instructions and use that to mask loaded values when code is executed speculatively.
- For ARM ISAs, we can use their recommended conditional move patterns along with the new barrier instruction.

## Jump table bounds checking

Cretonne doesn't generate jump tables yet, but it is expected to do so in the future to get good performance for `br_table` instructions.

The bounds check on a jump table load can be exploited in much the same way as a heap bounds check. The speculatively executed indirect branch would go to an address that was loaded from outside the jump table. This could potentially be a "speculative gadget" that could be used to leak data.

To mitigate this, we can make sure that jump table sizes are a power of two and use index masking to prevent speculative loads from accessing memory outside the jump table.

Independent from that, we may also have to use a "retpoline" instead of an indirect branch to avoid indirect branch predictor aliasing. The retpoline alone does not prevent the bounds check exploit.

## Table call bounds checking

WebAssembly indirect calls use a table of function pointers and trap if you attempt to call a function index outside the table. This bounds check can be exploited the same way as a jump table bounds check.

And we need to use retpolines for the indirect call too.

**Comment by jfoote** [comment]
That makes sense to me, thanks Jakob. 

This GH issue was meant to be an inquiry -- I'll close it for now. If you'd prefer to use it to track this as a work item feel free to re-open.

**Comment by stoklund** [comment]
Thanks, Jonathan. I'll leave it open to track remaining work.

**Comment by tedwin1113** [comment]
Suggested alternate usage scenario: detect only.

No idea how to build this (last time I used an assembler was in 1984 !;), but given these vulnerabilities may never be fixed, seems to me some detection-only tools able to identify live on-line sources of infection would be very helpful. Even more important, some means of detecting tainted package/module updates. 

As one who has extra non-critical-use machines running 24x7, I'd gladly invest time fitting some of them out with detection tools, even with at a performance penalty. Not sure what a 'gadget' is, but seems like it might be possible to plant some 'special-gadgets' capable of leaving breadcrumbs.(?) Also wonder if a 'whistleblower' exploit could be concocted, run-solo on a clean system while storing profile data, then, following each upgrade, detect interference by a potential unsub, by comparing profile changes. 

Thanks to all working this seriously fubar situation.
